server/src/main/java/org/elasticsearch/search/fetch/subphase/FetchFieldsPhase.java

```diff
@@ -10,48 +10,102 @@
-import org.elasticsearch.search.SearchHit;
+import org.elasticsearch.index.mapper.IdFieldMapper;
+import org.elasticsearch.index.mapper.IgnoredFieldMapper;
+import org.elasticsearch.index.mapper.LegacyTypeFieldMapper;
+import org.elasticsearch.index.mapper.MappedFieldType;
+import org.elasticsearch.index.mapper.RoutingFieldMapper;
+import org.elasticsearch.index.mapper.SourceFieldMapper;
+import org.elasticsearch.index.query.SearchExecutionContext;
 import org.elasticsearch.search.fetch.FetchContext;
 import org.elasticsearch.search.fetch.FetchSubPhase;
 import org.elasticsearch.search.fetch.FetchSubPhaseProcessor;
+import org.elasticsearch.search.fetch.StoredFieldsContext;
 import org.elasticsearch.search.fetch.StoredFieldsSpec;
 
 import java.io.IOException;
+import java.util.Collections;
+import java.util.Comparator;
+import java.util.List;
 import java.util.Map;
+import java.util.Set;
+import java.util.TreeSet;
 
 /**
  * A fetch sub-phase for high-level field retrieval. Given a list of fields, it
  * retrieves the field values through the relevant {@link org.elasticsearch.index.mapper.ValueFetcher}
  * and returns them as document fields.
  */
 public final class FetchFieldsPhase implements FetchSubPhase {
+
+    private static final List<FieldAndFormat> DEFAULT_METADATA_FIELDS = List.of(
+        new FieldAndFormat(IgnoredFieldMapper.NAME, null),
+        new FieldAndFormat(RoutingFieldMapper.NAME, null),
+        new FieldAndFormat(LegacyTypeFieldMapper.NAME, null)
+    );
+
     @Override
     public FetchSubPhaseProcessor getProcessor(FetchContext fetchContext) {
-        FetchFieldsContext fetchFieldsContext = fetchContext.fetchFieldsContext();
-        if (fetchFieldsContext == null) {
+        final FetchFieldsContext fetchFieldsContext = fetchContext.fetchFieldsContext();
+        final StoredFieldsContext storedFieldsContext = fetchContext.storedFieldsContext();
+
+        boolean fetchStoredFields = storedFieldsContext != null && storedFieldsContext.fetchFields();
+        if (fetchFieldsContext == null && fetchStoredFields == false) {
             return null;
         }
 
-        FieldFetcher fieldFetcher = FieldFetcher.create(fetchContext.getSearchExecutionContext(), fetchFieldsContext.fields());
+        final SearchExecutionContext searchExecutionContext = fetchContext.getSearchExecutionContext();
+        final FieldFetcher fieldFetcher = fetchFieldsContext == null ? null
+            : fetchFieldsContext.fields() == null ? null
+            : fetchFieldsContext.fields().isEmpty() ? null
+            : FieldFetcher.create(searchExecutionContext, fetchFieldsContext.fields());
 
+        final FieldFetcher metadataFieldFetcher;
+        if (storedFieldsContext != null
+            && storedFieldsContext.fieldNames() != null
+            && storedFieldsContext.fieldNames().isEmpty() == false) {
+            final Set<FieldAndFormat> metadataFields = new TreeSet<>(Comparator.comparing(fieldAndFormat -> fieldAndFormat.field));
```

Comment:
why not use a list directly given that order matters? You wanted to remove duplicates?

Reply:
When calling `getMatchingFieldNames` with a `*` it would return also fields like `_ignored`, `_routing` which would be added by default. At the end duplicates do not really cause an issue because a map used to add them to the document and the map eliminates duplicates because uses the name of the field as key. Note that this happens also in your branch...I verified it by actually running the code.

Reply:
sounds good, but with your comparator you change the order to be based on field name instead of insertion order? Maybe it is a good idea to remove duplicates, because otherwise we do end up fetching the same fields multiple times, although they get  de-duplicated later.

Reply:
Ok I can replace the `TreeSet` with the `HashSet`.